x-pack/docs/en/security/authentication/saml-guide.asciidoc

```diff
@@ -146,8 +146,8 @@ xpack.security.authc.realms.saml.saml1::
 
 order::
     You should define a unique order on each realm in your authentication chain.
-    It is recommended that the SAML realm be at the bottom of your authentication
-    chain (that is, that it has the _highest_ order).
+    Realms with a lower order are consulted first. Although not required, use of this setting is strongly recommended when you configure multiple realms. Defaults to Integer.MAX_VALUE.
+
```

Comment:
I feel like this actually changes the advice.
Previously we provided an explicit recommendation about what to do, but this proposal would just give general documentation about the `order`.

Reply:
Sorry, feel free to change it back. I just notice the definition in this doc is different from https://www.elastic.co/guide/en/elasticsearch/reference/7.8/security-settings.html#ref-realm-settings and when cross-referencing them, I was confused whether the value should it be high or low. The word `bottom` is throwing me off a bit, but your statement above `most commonly used realm with a higher priority (lower order)` makes more sense.

Reply:
After reading through this issue, I propose the following description. I think that "end of the chain" makes more sense than "bottom," and hopefully provides some clarity. @tvernum, thoughts?

>The priority of the realm within the realm chain. SAML realms are consulted in order starting at the end of the chain, with realms in the lowest position having the highest order. We recommend putting your most common realm at the end of your realm chain so that it is consulted first. This positioning can make authentication faster because additional realms are not consulted when authenticating.

If there are multiple SAML realms, then the order matters because we still consult the SAML realms in order, so having your most commonly used realm with a higher priority (lower order) will be a tiny bit more efficient (even if Kibana specifies the realm name, we still iterate through them until 1 of them authenticates successfully)

Even with a single SAML realm, having them at the end of the chain makes the other authentication methods faster because the SAML realm is never consulted.

Reply:
@tvernum, please let me know if you're 👍 to my proposed changes in the previous comment.

Reply:
@ywangd, can you please review my suggested change above?

Reply:
@lockewritesdocs I'd prefer we consistently make the usages of "high/low priority" and "high/low order (value)" instead of using either "bottom" or "end" or "position" (To me, they don't naturally suggest what comes first, while priority does). In my opinion, the confusing bit really is priorty and order are in reverse relationship and we should clearly state that. So something along the line of "Higher priority realms are consulted first and lower order value means higher priorty".

Reply:
Sorry, @lockewritesdocs I'm running about a week behind on my task list.
I agree with @ywangd that consistency would be good.

I guess there's the question of what we really want to try and say here. How much of the low level details are we trying to explain, vs just give the reader advice they can apply.

Speaking informally, my advice is:
- SAML realms should have an `order` that is numerically higher than other realms (by "other" realms, I mean native, LDAP, AD etc. Since SAML and OpenID Connect are so similar, it's hard to make any claim about what relative order they should have compared to each other).
- If you have multiple SAML realms (which is rare), then the most commonly used realm should have an `order` that is  numerically lower than the other SAML realms.

Neither of those pieces of advice are very important, it's mainly there because customers often like to be given advice what to do, rather than having to guess, and that's the best possible advice (even though it's of minor importance).

Trying to put that into a short statement, I would try something like:

> The order of the realm within the realm chain. Realms with a lower order are consulted first. It is recommended that password based realm (such as file, native, ldap and active_directory) have the lowest order, followed by SSO realms (such as SAML and OpenID Connect). If you have multiple realms of the same type, the more frequently used should typically be placed first (lower numerical order)

Reply:
Thank you @ywangd and @tvernum for your input! I like Tim's suggestion, and made some minor changes to it:
>The order of the realm within the realm chain. Realms with a lower order have
highest priority and are consulted first. We recommend giving password-based
realms such as file, native, LDAP, and Active Directory the lowest order (highest priority), followed by SSO realms such as SAML and OpenID Connect. If you have multiple realms of the same type, give the most frequently accessed
realm the lowest order to have it consulted first.

We should also update other mentions of `order`, such as in the [OpenID Connect](https://www.elastic.co/guide/en/elasticsearch/reference/current/oidc-guide-authentication.html) docs.

Reply:
@tvernum, please review my suggested changes based on your statement ☝️ and let me know if further changes are required. Thanks!

Reply:
LGTM, Thanks @lockewritesdocs